### ISSUE TITLE
Sometimes the pen doesn't draw

### ISSUE THREAD

**Issue by sclements** [LLK/scratch-flash#482]
http://scratch.mit.edu/discuss/topic/51605/

We should draw a tiny circle instead of a tiny diagonal line:
https://github.com/LLK/scratch-flash/blob/master/src/primitives/MotionAndPenPrims.as#L230

**Comment by sclements** [comment]
Production vs Fixed:
![image](https://cloud.githubusercontent.com/assets/874656/6985423/6bd106de-d9f0-11e4-868d-69bdce044959.png)

**Comment by TheLogFather** [comment]
Was just curious if this is really an efficient way to fix this...?
It'll work, of course, but is it a good idea to draw a whole circle at *every* pen down, when the more likely thing that's going to happen next is the x,y co-ords being moved to draw a line (which works fine without this fix), and when extra unnecessary rendering (of the circle) might lead to performance issues at the next refresh?
Would it be better to only bother drawing such a circle on "pen up" instead, but only if the pen has not moved (enough) since last "pen down"...? (Would need to cache the the original pen colour and size at "pen down", in order to get the right colour and size of the circle at "pen up", but that's not hard...)

**Comment by TheLogFather** [comment]
Another thought about this is to do with aliasing.
I don't know whether you've noticed, but at the moment there are some (usually) subtle effects due to this stroke that always happens at pen down - it can lead to a slight 'protrusion' at the start of a pen-drawn line, because there were *two* things drawn (the tiny stroke at "pen down" at the start of the line, then the actual line to the new position). Because of aliasing, these two things can add up to some subtly odd-looking behaviour where the edges of closed polygons are supposed to 'join up'.
My suggestion above would also remove such issues - it would only draw either a single line or a single circle.

**Comment by sclements** [comment]
@TheLogFather My choice of using a circle with a diameter of the pen width should behave almost exactly the same as before. We use round line caps and, theoretically, a very short line will be the same as a circle. For reference, here are the supported line styles and what they look like:
![image](https://cloud.githubusercontent.com/assets/874656/6989546/e5d2b7c4-da19-11e4-9308-539abdb5dd2a.png)

With the previous code sometimes nothing would draw. With the current code you will always start with a small circle, but the cap on the end of the line should draw exactly on top of the circle. We could go further to optimize by, as you said, only drawing the circle if the pen does not move. However, I know that this would be much more involved and probably have multiple edge cases as well. Regarding the visual quality, we shouldn't see a difference.

**Comment by griffpatch** [comment]
I think @LogFather has a point, drawing exactly the same circle over
another circle will create a visually larger slightly less circular circle
due to the semi translated anti alialised pixels being draw twice as heavy,
but only at the corner points. This might even effect some projects such as
my pen fill project? (But I haven't checked). But also, and equally as
important if not more so, does this impact performance? There are a lot of
own drawing programs or there that require things to go very fast and will
not see any improvement from this pen glitch. Can we bench mark before and
after?

With regard to the fix, would it not be the best case to only draw a circle
at pen up when the sprite has not moved since pen down?

**Comment by griffpatch** [comment]
In retrospect I think I just repeated exactly what @TheLogFather said  ;).
However implementation wise it may be easier (may), so add a penInvalidate
boolean which is set to true at penDown, and then is set to false when any
goto or change x or y is performed, but back to true on colour change. Then
on mouse up you only draw a circle if penInvalidate is true.

**Comment by TheLogFather** [comment]
"With the previous code sometimes nothing would draw."
Yes, I've been aware of that since writing InterXeptor a couple of years ago. That was when I found I needed to round to the nearest half pixel to be sure to get the dots to render for the stars...

"My choice of using a circle with a diameter of the pen width should behave almost exactly the same as before."
Yes, and I've been meaning to bring it up in this issue back when I first saw it appear - that the short stroke method (i.e. 'before') can cause this aliasing problem, and so it might be an idea to only draw it if required on pen up. The circle will likely behave in the same way.



However, I've also recently come to a better understanding of how drawing in Flash affects the refresh time (i.e. *after* script has finished executing. See https://scratch.mit.edu/projects/51157420/
Drawing extra things (unnecessarily) can fairly significantly reduce performance under certain circumstances. Hence my thought that drawing only on "pen up", if required, should help.

Can you be more specific about what edge cases you have in mind? AFAICT, a dot only gets drawn at "pen down", and changing the pen size/colour/shade/etc. has no effect if it doesn't move. That means you only need to save these attributes at "pen down", and use them to draw the circle at "pen up" if the pen has not moved, to make it behave as it currently does.

**Comment by TheLogFather** [comment]
"but back to true on colour change" - not sure that's needed... changing the pen colour doesn't change the colour of the dot unless the pen subsequently moves.

**Comment by griffpatch** [comment]
Hmm... Indeed. (although thinking about it perhaps it should since the pen
is down). However currently it doesn't so yeah it wouldn't need to change
anything on pen colour change.

**Comment by TheLogFather** [comment]
OK, I've managed to come up with a couple of scenarios where my suggestion can behave in a way that's different to the current behaviour:

1. If a project does more than 0.5s of non-refresh work right after "pen down", and before it does "pen up", then the forced refresh currently shows the dot that's drawn at "pen down", but it would not if drawn only at "pen up". (The dot would eventually get drawn, though, at the next refresh after "pen up".)

2. Suppose one sprite/clone does "pen down", but doesn't move the pen, and then another sprite/clone draws something which would overlap the dot, and only after that does the first sprite do "pen up".
The order of drawing would be changed - the dot is currently drawn before the line (as expected), but if it's drawn only at "pen up" then it would get drawn after the line.

I think sclements is correct in saying that it is "more involved"...

Here's a suggestion, though:
How about keeping a global list of "pen down"s (along with their pen attributes - position, size & colour), and have each sprite/clone able to remove items it 'owns' from that list when it does "pen up" or moves its pen?
I think flushing this list (i.e. draw any dots it contains) under two circumstances should ensure both of the above cases get dealt with correctly: firstly, just before allowing a refresh (this deals with case 1, and case 2 if the "pen up" happens after the next refresh); secondly, just before a sprite/clone moves its pen or stamps or does "pen up" - though, if it is moving its pen, then it should first remove any items in the list that it owns (this deals with case 2 if the "pen up" manages to happen before the next refresh).

Any other cases not covered here?

**Comment by griffpatch** [comment]
That sounds like an even more involved fix, but if you were going to the
bother of implementing the sorting of drawing points then it would
naturally lead to the ability to draw filed polygons *grin*... And that
would be transformational with regard to pen based projects! All you'd need
is a new set polygon fill colour block and when pen up is reached draw the
full polygon filled :)

**Comment by TheLogFather** [comment]
@griffpatch: not sure I follow what you mean - why would it lead to the ability to draw filled polygons? (Though it would certainly be nice to have that capability built-in...)

@sclements:
I realised I had a way to get some idea of the performance impact of this extra unnecessary dot using my triangle filler framework, https://scratch.mit.edu/projects/55095796

I can reduce the effect of the dot on Flash's screen refresh time by ensuring that the pen has size 1 before pen down, and then only set the pen to the actual width I want for the line just before moving it. That means the initial dot is only drawn with size 1, but the subsequent line gets drawn at the desired width.

Doing this boosts the performance by about 5%, so it's not huge, but it's also not insignificant (though it is also having to process those extra "pen size" blocks, which no doubt adds a teeny bit extra script execution time into the mix).

**Comment by griffpatch** [comment]
Hey, what I was getting at was that if the effort say out into adding the
recording of drawn points which were then only drawn on pen up then it
would only be a small step to then add a new block to allow the pen up to
use those recorded points as a polygon and draw that instead? :) wishful
thinking of course!

**Comment by 2jour** [comment]
my port it works on mac 10.10 and chrome 41

but on linux (sorry don't know distro :( )  ff 37 pen behaves as before 

https://saucelabs.com/tests/cd89994f3ac44469b85c5a2e07e0835e

**Comment by 2jour** [comment]
@griffpatch do you have a sample project that i can take look at for testing :) ?

**Comment by griffpatch** [comment]
Well, you could use this one https://scratch.mit.edu/projects/14318033/
which requires exact placement of own lines? But for speed tests of ask
@DadOfMrLog

**Comment by TheLogFather** [comment]
Has this fix gone 'live' now in 434b?

Here's a speed-test project that demonstrates the performance hit from the dot drawn at pen down:
https://scratch.mit.edu/projects/56234668/

**Comment by 2jour** [comment]
Thank you @TheLogFather and @griffpatch :)
